src/tools/w3dviewer/w3dviewdoc.cpp

```diff
@@ -0,0 +1,957 @@
+                                view->m_camera->Set_Transform(tm);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+void CW3DViewDoc::AnimateCamera(bool animate)
+{
+    m_animateCamera = animate;
+
+    if (!animate) {
+        CMainFrame *frame = static_cast<CMainFrame *>(AfxGetMainWnd());
+        frame->SendMessage(WM_COMMAND, ID_CAMERA_RESET);
+    }
+}
+
+void CW3DViewDoc::EnableFog(bool enable)
+{
+    if (m_scene != nullptr) {
+        m_scene->Set_Fog_Enable(enable);
+        m_fogEnabled = enable;
+    }
+}
+
+void CW3DViewDoc::OnStep(int step)
+{
+    if (m_model != nullptr) {
+        if (m_animation != nullptr) {
+            int frames = m_animation->Get_Num_Frames();
+            m_frameCount += step;
+
+            if (m_frameCount < frames) {
+                if (m_frameCount < 0.0f) {
+                    m_frameCount = frames - 1;
+                }
+            } else {
+                m_frameCount = 0.0f;
+                m_time = 0.0f;
+            }
+
+            float speed = GetCurrentGraphicView()->m_animationSpeed;
+            CMainFrame *frame = static_cast<CMainFrame *>(AfxGetMainWnd());
+            float rate = m_animation->Get_Frame_Rate();
+            frame->UpdateFrameCount(m_frameCount, frames - 1, speed * rate);
+
+            if (m_blendFrames) {
+                m_model->Set_Animation(m_animation, m_frameCount);
+            } else {
+                m_model->Set_Animation(m_animation, (int)m_frameCount);
+            }
+
+            if (m_animateCamera) {
+                if (m_model != nullptr) {
+                    Matrix3D tm(true);
+
+                    if (GetCameraBoneTransform(m_model, &tm)) {
+                        Matrix3D m;
+                        m[0].X = 0.0f;
+                        m[0].Y = 0.0f;
+                        m[0].Z = -1.0f;
+                        m[0].W = 0.0f;
+                        m[1].X = -1.0f;
+                        m[1].Y = 0.0f;
+                        m[1].Z = 0.0f;
+                        m[1].W = 0.0f;
+                        m[2].X = 0.0f;
+                        m[2].Y = 1.0f;
+                        m[2].Z = 0.0f;
+                        m[2].W = 0.0f;
+                        tm = tm * m;
+                        CMainFrame *frame = static_cast<CMainFrame *>(AfxGetMainWnd());
+
+                        if (frame != nullptr) {
+                            CGraphicView *view = static_cast<CGraphicView *>(frame->m_splitter.GetPane(0, 1));
+
+                            if (view != nullptr) {
+                                view->m_camera->Set_Transform(tm);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+void CW3DViewDoc::ToggleAlternateMaterials(RenderObjClass *robj)
+{
+    if (robj == nullptr) {
+        robj = m_model;

```

Comment:
What if `m_model` is null? Can put an assert if it is not meant to be null here ever by design.

Reply:
fixed.